File: app/src/main/java/fr/nihilus/music/ui/songs/SongListFragment.java
/*
 * Copyright 2017 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.nihilus.music.ui.songs;

import android.arch.lifecycle.ViewModelProviders;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.os.ResultReceiver;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.media.MediaBrowserCompat.MediaItem;
import android.support.v4.media.MediaBrowserCompat.SubscriptionCallback;
import android.support.v4.media.session.PlaybackStateCompat;
import android.support.v4.view.ViewCompat;
import android.support.v4.widget.ContentLoadingProgressBar;
import android.support.v7.content.res.AppCompatResources;
import android.support.v7.widget.SearchView;
import android.util.Log;
import android.util.SparseBooleanArray;
import android.view.ActionMode;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AbsListView.MultiChoiceModeListener;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

import dagger.android.support.AndroidSupportInjection;
import fr.nihilus.music.Constants;
import fr.nihilus.music.R;
import fr.nihilus.music.command.DeleteTracksCommand;
import fr.nihilus.music.command.MediaSessionCommand;
import fr.nihilus.music.library.BrowserViewModel;
import fr.nihilus.music.utils.ConfirmDialogFragment;
import fr.nihilus.music.utils.MediaID;

import static fr.nihilus.music.utils.MediaID.ID_MUSIC;

public class SongListFragment extends Fragment implements AdapterView.OnItemClickListener,
        AdapterView.OnItemLongClickListener {

    private static final String TAG = "SongListFragment";
    private static final String KEY_SCROLL = "ScrollY";
    private static final int REQUEST_CODE_DELETE_TRACKS = 21;

    private ListView mListView;
    private View mListContainer;
    private SongAdapter mAdapter;
    private ContentLoadingProgressBar mProgressBar;

    private BrowserViewModel mViewModel;
    private final SongListActionMode mActionMode = new SongListActionMode();

    public static SongListFragment newInstance() {
        Bundle args = new Bundle(1);
        args.putInt(Constants.FRAGMENT_ID, R.id.action_all);
        SongListFragment fragment = new SongListFragment();
        fragment.setArguments(args);
        return fragment;
    }

    private final SubscriptionCallback mCallback = new SubscriptionCallback() {
        @Override
        public void onChildrenLoaded(@NonNull String parentId, @NonNull List<MediaItem> items) {
            Log.d(TAG, "onChildrenLoaded: loaded " + items.size() + " from " + parentId);
            mAdapter.updateItems(items);
            mProgressBar.hide();
            mListContainer.setVisibility(View.VISIBLE);
        }
    };

    @Override
    public void onAttach(Context context) {
        AndroidSupportInjection.inject(this);
        super.onAttach(context);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
        mAdapter = new SongAdapter(this);
    }

    @Override
    public void onCreateOptionsMenu(Menu menu, MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_songlist, menu);
        final SearchView searchView = (SearchView) menu.findItem(R.id.action_search).getActionView();
        // TODO Search and filtering features
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_songs, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        mListContainer = view.findViewById(R.id.list_container);

        mListView = view.findViewById(R.id.list);
        setupListHeader();
        mListView.setAdapter(mAdapter);
        mListView.setOnItemClickListener(this);
        mListView.setEmptyView(view.findViewById(android.R.id.empty));
        mListView.setChoiceMode(ListView.CHOICE_MODE_MULTIPLE_MODAL);
        mListView.setMultiChoiceModeListener(mActionMode);
        ViewCompat.setNestedScrollingEnabled(mListView, true);

        mProgressBar = view.findViewById(android.R.id.progress);

        mListContainer.setVisibility(View.GONE);
        if (savedInstanceState == null) {
            mProgressBar.show();
        } else {
            mListView.setSelectionFromTop(savedInstanceState.getInt(KEY_SCROLL), 0);
        }
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mViewModel = ViewModelProviders.of(getActivity()).get(BrowserViewModel.class);
    }

    @Override
    public void onStart() {
        super.onStart();
        getActivity().setTitle(R.string.all_music);
        mViewModel.subscribe(MediaID.ID_MUSIC, mCallback);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        if (isVisible()) {
            outState.putInt(KEY_SCROLL, mListView.getFirstVisiblePosition());
        }
    }

    @Override
    public void onStop() {
        mViewModel.unsubscribe(MediaID.ID_MUSIC);
        super.onStop();
    }

    @Override
    public void onDestroyView() {
        mListView = null;
        mProgressBar = null;
        mListContainer = null;
        super.onDestroyView();
    }

    private void setupListHeader() {
        final Context context = getContext();
        final LayoutInflater inflater = LayoutInflater.from(context);
        View headerView = inflater.inflate(R.layout.random_button, mListView, false);
        Drawable icRandom = AppCompatResources.getDrawable(context, R.drawable.ic_shuffle_primary);
        ((TextView) headerView.findViewById(R.id.text)).setCompoundDrawables(icRandom, null, null, null);
        mListView.addHeaderView(headerView);
    }

    @Override
    public void onItemClick(AdapterView<?> listView, View view, int position, long id) {
		if (position == 0) {
			mViewModel.setShuffleMode(PlaybackStateCompat.SHUFFLE_MODE_ALL);
			mViewModel.playFromMediaId(MediaID.createMediaID(null, ID_MUSIC));
		} else {
			// Offset the position as the header is considered at position 0
			MediaItem clickedItem = mAdapter.getItem(position - 1);
			mViewModel.playFromMediaId(clickedItem.getMediaId());
		}
    }

    @Override
    public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
        return false;
    }

    private void showDeleteDialog() {
        String dialogMessage = getResources().getQuantityString(R.plurals.delete_dialog_message,
                mListView.getCheckedItemCount(), mListView.getCheckedItemCount());

        ConfirmDialogFragment confirm = ConfirmDialogFragment.newInstance(this, 21,
                getString(R.string.delete_dialog_title), dialogMessage,
                R.string.action_delete, R.string.cancel, 0);
        confirm.show(getFragmentManager(), null);
    }

    private void deleteSelectedTracks() {
        int index = 0;
        SparseBooleanArray checked = mListView.getCheckedItemPositions();
        final long[] toDelete = new long[mListView.getCheckedItemCount()];
        for (int i = 0; i < checked.size(); i++) {
            if (checked.valueAt(i)) {
                int pos = checked.keyAt(i);
                toDelete[index++] = mAdapter.getItemId(pos - 1);
            }
        }

        Bundle params = new Bundle(1);
        params.putLongArray(DeleteTracksCommand.PARAM_TRACK_IDS, toDelete);
        mViewModel.sendCommand(DeleteTracksCommand.CMD_NAME, params, new ResultReceiver(new Handler()) {

            @Override
            protected void onReceiveResult(int resultCode, Bundle resultData) {
                View rootView = getView();
                if (resultCode == MediaSessionCommand.CODE_SUCCESS && rootView != null) {
                    String message = getResources()
                            .getQuantityString(R.plurals.deleted_songs_confirmation, toDelete.length);
                    Toast.makeText(getContext(), message, Toast.LENGTH_LONG).show();
                }
            }
        });
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == REQUEST_CODE_DELETE_TRACKS
                && resultCode == DialogInterface.BUTTON_POSITIVE) {
            deleteSelectedTracks();
            mActionMode.finish();
        }
    }

    /**
     * An ActionMode that handles multiple item selection inside the song ListView.
     */
    private class SongListActionMode implements MultiChoiceModeListener {
        ActionMode mActionMode;

        @Override
        public void onItemCheckedStateChanged(ActionMode mode, int position, long id, boolean checked) {
            mode.setTitle(String.valueOf(mListView.getCheckedItemCount()));
        }

        @Override
        public boolean onCreateActionMode(ActionMode mode, Menu menu) {
            mode.getMenuInflater().inflate(R.menu.actionmode_songlist, menu);
            mActionMode = mode;
            return true;
        }

        @Override
        public boolean onPrepareActionMode(ActionMode mode, Menu menu) {
            return false;
        }

        @Override
        public boolean onActionItemClicked(ActionMode mode, MenuItem item) {
            switch (item.getItemId()) {
                case R.id.action_delete:
                    showDeleteDialog();
                    return true;
                case R.id.action_playlist:
                    // TODO Prepare a playlist with the selected items
                    mode.finish();
                    return true;
                default:
                    return false;
            }
        }

        @Override
        public void onDestroyActionMode(ActionMode mode) {
            mActionMode = null;
            // By default, deselect items
        }

        void finish() {
            if (mActionMode != null) {
                mActionMode.finish();
            }
        }
    }
}

File: app/src/main/java/fr/nihilus/music/ui/songs/SongListFragment.kt
/*
 * Copyright 2018 Thibault Seisel
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package fr.nihilus.music.ui.songs

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.support.v4.media.MediaBrowserCompat
import android.view.*
import android.widget.AbsListView.MultiChoiceModeListener
import android.widget.AdapterView
import android.widget.ListView
import android.widget.Toast
import fr.nihilus.music.BaseFragment
import fr.nihilus.music.R
import fr.nihilus.music.client.FRAGMENT_ID
import fr.nihilus.music.isVisible
import fr.nihilus.music.library.songs.SongListViewModel
import fr.nihilus.music.media.CATEGORY_MUSIC
import fr.nihilus.music.ui.playlist.AddToPlaylistDialog
import fr.nihilus.music.ui.playlist.NewPlaylistDialog
import fr.nihilus.music.utils.ConfirmDialogFragment
import fr.nihilus.music.utils.LoadRequest
import fr.nihilus.music.utils.ProgressTimeLatch
import kotlinx.android.synthetic.main.fragment_songs.*

class SongListFragment : BaseFragment() {

    private val multiSelectMode = SongListActionMode()

    private lateinit var songAdapter: SongAdapter
    private lateinit var viewModel: SongListViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
        songAdapter = SongAdapter(this)
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_songlist, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem) = when (item.itemId) {
        R.id.action_play_shuffled -> {
            viewModel.playAllShuffled()
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View = inflater.inflate(R.layout.fragment_songs, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProviders.of(this, viewModelFactory).get(SongListViewModel::class.java)
        viewModel.loadChildrenOf(CATEGORY_MUSIC)

        val progressBarLatch = ProgressTimeLatch { shouldShowProgress ->
            progress_indicator.isVisible = shouldShowProgress
            songs_listview.isVisible = !shouldShowProgress
        }

        songs_listview.run {
            adapter = songAdapter
            choiceMode = ListView.CHOICE_MODE_MULTIPLE_MODAL
            setMultiChoiceModeListener(multiSelectMode)

            onItemClickListener = AdapterView.OnItemClickListener { _, _, position, _ ->
                viewModel.onSongSelected(songAdapter.getItem(position))
            }
        }

        viewModel.items.observe(this, Observer { itemRequest ->
            when (itemRequest) {
                is LoadRequest.Pending -> progressBarLatch.isRefreshing = true
                is LoadRequest.Success -> {
                    progressBarLatch.isRefreshing = false
                    songAdapter.updateItems(itemRequest.data)
                    group_empty_view.isVisible = itemRequest.data.isEmpty()
                }
                is LoadRequest.Error -> {
                    progressBarLatch.isRefreshing = false
                }
            }
        })

        viewModel.toastMessage.observe(this, Observer { toastMessageEvent ->
            toastMessageEvent?.handle { message ->
                Toast.makeText(context, message, Toast.LENGTH_LONG).show()
            }
        })
    }

    override fun onStart() {
        super.onStart()
        activity!!.setTitle(R.string.all_music)
    }

    private fun showDeleteDialog() {
        val checkedItemCount = songs_listview.checkedItemCount
        val dialogMessage = resources.getQuantityString(
            R.plurals.delete_dialog_message,
            checkedItemCount, checkedItemCount
        )

        val confirm = ConfirmDialogFragment.newInstance(
            this, R.id.request_delete_tracks,
            getString(R.string.delete_dialog_title), dialogMessage,
            R.string.action_delete, R.string.cancel, 0
        )
        confirm.show(fragmentManager, null)
    }

    private fun deleteSelectedTracks() {
        val checkedSongPositions = songs_listview.checkedItemPositions
        val songsToDelete = mutableListOf<MediaBrowserCompat.MediaItem>()
        for (pos in 0 until checkedSongPositions.size()) {
            if (checkedSongPositions[pos]) {
                songsToDelete += songAdapter.getItem(pos)
            }
        }

        viewModel.deleteSongs(songsToDelete)
    }

    private fun openPlaylistChooserDialog() {
        val checkedIds = songs_listview.checkedItemIds
        val dialog = AddToPlaylistDialog.newInstance(this, R.id.request_add_to_playlist, checkedIds)
        dialog.show(fragmentManager, AddToPlaylistDialog.TAG)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {

        // User has confirmed his intent to delete track(s)
        if (requestCode == R.id.request_delete_tracks && resultCode == DialogInterface.BUTTON_POSITIVE) {
            deleteSelectedTracks()
            multiSelectMode.finish()
        }

        else if (requestCode == R.id.request_add_to_playlist) {

            when (resultCode) {
                R.id.abc_result_success -> {
                    // Track(s) have been added to a playlist
                    if (data != null) {
                        // Display a confirmation message as a Toast
                        val count = data.getIntExtra(AddToPlaylistDialog.RESULT_TRACK_COUNT, 0)
                        val title = data.getStringExtra(AddToPlaylistDialog.RESULT_PLAYLIST_TITLE)

                        val userMessage = resources.getQuantityString(
                            R.plurals.tracks_added_to_playlist, count, count, title)
                        Toast.makeText(context, userMessage, Toast.LENGTH_SHORT).show()
                    }

                    // Finish action mode to deselect all items
                    multiSelectMode.finish()
                }

                R.id.abc_error_playlist_already_exists -> {
                    // Failed to insert to a playlist due to its name being already taken
                    val title = data!!.getStringExtra(NewPlaylistDialog.RESULT_TAKEN_PLAYLIST_TITLE)

                    val userMessage = getString(R.string.error_playlist_title_taken, title)
                    Toast.makeText(context, userMessage, Toast.LENGTH_LONG).show()
                }
            }
        }
    }

    /**
     * An ActionMode that handles multiple item selection inside the song ListView.
     */
    private inner class SongListActionMode : MultiChoiceModeListener {
        private var actionMode: ActionMode? = null

        override fun onItemCheckedStateChanged(mode: ActionMode, position: Int,
                                               id: Long, checked: Boolean) {
            mode.title = songs_listview.checkedItemCount.toString()
        }

        override fun onCreateActionMode(mode: ActionMode, menu: Menu): Boolean {
            mode.menuInflater.inflate(R.menu.actionmode_songlist, menu)
            actionMode = mode
            return true
        }

        override fun onPrepareActionMode(mode: ActionMode, menu: Menu) = false

        override fun onActionItemClicked(mode: ActionMode, item: MenuItem) = when (item.itemId) {
            R.id.action_delete -> {
                showDeleteDialog()
                true
            }
            R.id.action_playlist -> {
                openPlaylistChooserDialog()
                true
            }
            else -> false
        }

        override fun onDestroyActionMode(mode: ActionMode) {
            // Items are deselected by default when action mode is destroyed
            actionMode = null
        }

        fun finish() {
            actionMode?.finish()
        }
    }

    companion object Factory {
        fun newInstance(): SongListFragment {
            val args = Bundle(1)
            args.putInt(FRAGMENT_ID, R.id.action_all)
            val fragment = SongListFragment()
            fragment.arguments = args
            return fragment
        }
    }
}
